File: examples/kotlin/apps/agent/mcp_http.kt
package app.protocol.examples.kotlin.apps.agent

import app.protocol.examples.kotlin.core.shared.AppMcpHttpExchange
import app.protocol.examples.kotlin.core.shared.AppMcpHttpResponse
import app.protocol.examples.kotlin.core.shared.AppMcpProtocolError
import app.protocol.examples.kotlin.core.shared.AppMcpRequestContext
import app.protocol.examples.kotlin.core.shared.AppMcpServer
import app.protocol.examples.kotlin.core.shared.BaseAppMcpHttpAdapter
import kotlinx.serialization.encodeToString
import kotlinx.serialization.json.JsonArray
import kotlinx.serialization.json.JsonElement
import kotlinx.serialization.json.JsonObject
import kotlin.random.Random

private fun httpSessionId(): String = "mcp_http_${Random.nextLong(0, Long.MAX_VALUE).toString(16)}"

private fun httpJsonRpcFailure(
    id: JsonElement?,
    code: Int,
    message: String,
    data: Any? = null,
): JsonElement =
    encodeAgentValue(
        mapOf(
            "jsonrpc" to "2.0",
            "id" to id,
            "error" to mapOf(
                "code" to code,
                "message" to message,
                "data" to data,
            ),
        ),
    )

private fun isJsonRpcMessage(value: JsonElement): Boolean =
    value is JsonObject &&
        value["jsonrpc"]?.stringValueOrNull() == "2.0" &&
        value["method"]?.stringValueOrNull() != null

private fun toJsonResponse(
    statusCode: Int,
    bodyText: String? = null,
    headers: Map<String, String> = emptyMap(),
): AppMcpHttpResponse =
    AppMcpHttpResponse(
        statusCode = statusCode,
        headers =
            mapOf("cache-control" to "no-store") +
                if (bodyText != null) {
                    mapOf("content-type" to "application/json; charset=utf-8")
                } else {
                    emptyMap()
                } + headers,
        bodyText = bodyText,
    )

class StreamableHttpAppMcpAdapter : BaseAppMcpHttpAdapter() {
    override val transport: String = "streamable-http"
    override val endpointPath: String = "/mcp"

    override suspend fun handle(
        exchange: AppMcpHttpExchange,
        server: AppMcpServer,
    ): AppMcpHttpResponse? {
        if (exchange.path != endpointPath) {
            return null
        }

        val method = exchange.method.uppercase()
        if (method == "GET" || method == "DELETE") {
            return AppMcpHttpResponse(
                statusCode = 405,
                headers = mapOf("allow" to "POST", "cache-control" to "no-store"),
            )
        }

        if (method != "POST") {
            return AppMcpHttpResponse(
                statusCode = 405,
                headers = mapOf("allow" to "GET,POST,DELETE", "cache-control" to "no-store"),
            )
        }

        val bodyText = exchange.bodyText?.trim()
        if (bodyText.isNullOrBlank()) {
            return toJsonResponse(
                400,
                agentJson.encodeToString(JsonElement.serializer(), httpJsonRpcFailure(null, -32600, "Missing JSON-RPC payload.")),
            )
        }

        val parsed =
            try {
                agentJson.parseToJsonElement(bodyText)
            } catch (error: Throwable) {
                return toJsonResponse(
                    400,
                    agentJson.encodeToString(
                        JsonElement.serializer(),
                        httpJsonRpcFailure(null, -32700, "Invalid JSON-RPC payload.", error.message),
                    ),
                )
            }

        if (parsed is JsonArray && parsed.isEmpty()) {
            return toJsonResponse(
                400,
                agentJson.encodeToString(
                    JsonElement.serializer(),
                    httpJsonRpcFailure(null, -32600, "Empty JSON-RPC batch payload is invalid."),
                ),
            )
        }

        val messages = if (parsed is JsonArray) parsed else listOf(parsed)
        val responses = mutableListOf<JsonElement>()

        for (message in messages) {
            if (!isJsonRpcMessage(message)) {
                responses += httpJsonRpcFailure(null, -32600, "Invalid JSON-RPC request shape.")
                continue
            }

            val request = message as JsonObject
            val id = request["id"]
            if (id == null) {
                continue
            }

            try {
                responses += handleRequest(request, server)
            } catch (error: AppMcpProtocolError) {
                responses += httpJsonRpcFailure(id, error.code, error.message, error.data)
            } catch (error: Throwable) {
                responses += httpJsonRpcFailure(
                    id,
                    -32603,
                    "Internal MCP server error.",
                    mapOf("message" to (error.message ?: "Unknown error")),
                )
            }
        }

        if (responses.isEmpty()) {
            return AppMcpHttpResponse(
                statusCode = 202,
                headers = mapOf("cache-control" to "no-store"),
            )
        }

        val payload = if (responses.size == 1) responses.first() else JsonArray(responses)
        return toJsonResponse(
            200,
            agentJson.encodeToString(JsonElement.serializer(), payload),
        )
    }

    private suspend fun handleRequest(
        request: JsonObject,
        server: AppMcpServer,
    ): JsonElement {
        val id = request["id"]
        val method = request["method"]?.stringValueOrNull().orEmpty()
        val params = request["params"]
        val context =
            AppMcpRequestContext(
                transport = transport,
                requestId = id?.stringValueOrNull() ?: id?.longValueOrNull(),
                sessionId = httpSessionId(),
            )

        val result: Any =
            when (method) {
                "initialize" -> {
                    val typedParams = parseInitializeParams(params as? JsonObject)
                    server.initialize(typedParams, context)
                }
                "ping" -> emptyMap<String, String>()
                "tools/list" -> mapOf("tools" to server.listTools(context))
                "resources/list" -> mapOf("resources" to server.listResources(context))
                "resources/read" -> {
                    val uri = (params as? JsonObject)?.get("uri")?.stringValueOrNull()
                        ?: throw AppMcpProtocolError(-32602, "MCP resources/read requires a string resource uri.")
                    server.readResource(uri, context)
                }
                "tools/call" -> {
                    val name = (params as? JsonObject)?.get("name")?.stringValueOrNull()
                        ?: throw AppMcpProtocolError(-32602, "MCP tools/call requires a string tool name.")
                    val arguments = (params as? JsonObject)?.get("arguments")?.toPlainValue()
                    server.callTool(name, arguments, context)
                }
                else -> throw AppMcpProtocolError(-32601, "MCP method $method is not implemented by this server.")
            }

        return encodeAgentValue(
            mapOf(
                "jsonrpc" to "2.0",
                "id" to id,
                "result" to result,
            ),
        )
    }
}

File: examples/go/apps/agent/mcp_http.go
package agent

import (
	"encoding/json"
	"fmt"
	"strings"

	"app-protocol/examples/go/core/shared"
)

type StreamableHTTPAdapter struct{}

type jsonRPCRequest struct {
	JSONRPC string `json:"jsonrpc"`
	ID      any    `json:"id,omitempty"`
	Method  string `json:"method"`
	Params  any    `json:"params,omitempty"`
}

type jsonRPCResponse struct {
	JSONRPC string        `json:"jsonrpc"`
	ID      any           `json:"id"`
	Result  any           `json:"result,omitempty"`
	Error   *jsonRPCError `json:"error,omitempty"`
}

type jsonRPCError struct {
	Code    int    `json:"code"`
	Message string `json:"message"`
	Data    any    `json:"data,omitempty"`
}

func NewStreamableHTTPAdapter() *StreamableHTTPAdapter {
	return &StreamableHTTPAdapter{}
}

func (a *StreamableHTTPAdapter) Transport() string {
	return "streamable-http"
}

func (a *StreamableHTTPAdapter) EndpointPath() string {
	return "/mcp"
}

func (a *StreamableHTTPAdapter) Handle(exchange shared.AppMcpHTTPExchange, server shared.AppMcpServer) (*shared.AppMcpHTTPResponse, error) {
	if exchange.Path != a.EndpointPath() {
		return nil, nil
	}

	switch exchange.Method {
	case "GET", "DELETE":
		return &shared.AppMcpHTTPResponse{
			StatusCode: 405,
			Headers: map[string]string{
				"allow":         "POST",
				"cache-control": "no-store",
			},
		}, nil
	case "POST":
	default:
		return &shared.AppMcpHTTPResponse{
			StatusCode: 405,
			Headers: map[string]string{
				"allow":         "GET,POST,DELETE",
				"cache-control": "no-store",
			},
		}, nil
	}

	if stringsTrim(exchange.BodyText) == "" {
		return toJSONResponse(400, failureResponse(nil, -32600, "Missing JSON-RPC payload.", nil)), nil
	}

	var parsed any
	if err := json.Unmarshal([]byte(exchange.BodyText), &parsed); err != nil {
		return toJSONResponse(400, failureResponse(nil, -32700, "Invalid JSON-RPC payload.", err.Error())), nil
	}

	if batch, ok := parsed.([]any); ok && len(batch) == 0 {
		return toJSONResponse(400, failureResponse(nil, -32600, "Empty JSON-RPC batch payload is invalid.", nil)), nil
	}

	messages := []any{parsed}
	if batch, ok := parsed.([]any); ok {
		messages = batch
	}

	responses := []jsonRPCResponse{}
	for _, message := range messages {
		request, hasID, ok := parseJSONRPCMessage(message)
		if !ok {
			responses = append(responses, failureResponse(nil, -32600, "Invalid JSON-RPC request shape.", nil))
			continue
		}
		if !hasID {
			continue
		}

		response := handleJSONRPCRequest(request, server, &shared.AppMcpRequestContext{
			Transport: "streamable-http",
		})
		responses = append(responses, response)
	}

	if len(responses) == 0 {
		return &shared.AppMcpHTTPResponse{
			StatusCode: 202,
			Headers: map[string]string{
				"cache-control": "no-store",
			},
		}, nil
	}

	var body []byte
	var err error
	if len(responses) == 1 {
		body, err = json.Marshal(responses[0])
	} else {
		body, err = json.Marshal(responses)
	}
	if err != nil {
		return nil, err
	}

	return &shared.AppMcpHTTPResponse{
		StatusCode: 200,
		Headers: map[string]string{
			"cache-control": "no-store",
			"content-type":  "application/json; charset=utf-8",
		},
		BodyText: string(body),
	}, nil
}

func parseJSONRPCMessage(value any) (jsonRPCRequest, bool, bool) {
	record, ok := value.(map[string]any)
	if !ok {
		return jsonRPCRequest{}, false, false
	}
	if record["jsonrpc"] != "2.0" {
		return jsonRPCRequest{}, false, false
	}

	method, ok := record["method"].(string)
	if !ok || method == "" {
		return jsonRPCRequest{}, false, false
	}

	request := jsonRPCRequest{
		JSONRPC: "2.0",
		Method:  method,
		Params:  record["params"],
	}

	if id, exists := record["id"]; exists {
		request.ID = id
		return request, true, true
	}

	return request, false, true
}

func handleJSONRPCRequest(request jsonRPCRequest, server shared.AppMcpServer, parent *shared.AppMcpRequestContext) jsonRPCResponse {
	switch request.Method {
	case "initialize":
		params := &shared.AppMcpInitializeParams{}
		if request.Params != nil {
			encoded, _ := json.Marshal(request.Params)
			_ = json.Unmarshal(encoded, params)
		}
		result, err := server.Initialize(params, parent)
		if err != nil {
			return responseFromError(request.ID, err)
		}
		return successResponse(request.ID, result)
	case "ping":
		return successResponse(request.ID, map[string]any{})
	case "tools/list":
		result, err := server.ListTools(parent)
		if err != nil {
			return responseFromError(request.ID, err)
		}
		return successResponse(request.ID, map[string]any{"tools": result})
	case "resources/list":
		result, err := server.ListResources(parent)
		if err != nil {
			return responseFromError(request.ID, err)
		}
		return successResponse(request.ID, map[string]any{"resources": result})
	case "resources/read":
		record, _ := request.Params.(map[string]any)
		uri, _ := record["uri"].(string)
		if uri == "" {
			return failureResponse(request.ID, -32602, "MCP resources/read requires a string resource uri.", nil)
		}
		result, err := server.ReadResource(uri, parent)
		if err != nil {
			return responseFromError(request.ID, err)
		}
		return successResponse(request.ID, result)
	case "tools/call":
		record, _ := request.Params.(map[string]any)
		name, _ := record["name"].(string)
		if name == "" {
			return failureResponse(request.ID, -32602, "MCP tools/call requires a string tool name.", nil)
		}
		result, err := server.CallTool(name, record["arguments"], parent)
		if err != nil {
			return responseFromError(request.ID, err)
		}
		return successResponse(request.ID, result)
	default:
		return failureResponse(request.ID, -32601, fmt.Sprintf("Unsupported MCP method %s.", request.Method), nil)
	}
}

func successResponse(id any, result any) jsonRPCResponse {
	return jsonRPCResponse{
		JSONRPC: "2.0",
		ID:      id,
		Result:  result,
	}
}

func failureResponse(id any, code int, message string, data any) jsonRPCResponse {
	return jsonRPCResponse{
		JSONRPC: "2.0",
		ID:      id,
		Error: &jsonRPCError{
			Code:    code,
			Message: message,
			Data:    data,
		},
	}
}

func responseFromError(id any, err error) jsonRPCResponse {
	if protocolErr, ok := err.(*shared.AppMcpProtocolError); ok {
		return failureResponse(id, protocolErr.Code, protocolErr.Message, protocolErr.Data)
	}

	return failureResponse(id, -32603, "Internal MCP server error.", map[string]any{
		"message": err.Error(),
	})
}

func toJSONResponse(statusCode int, body jsonRPCResponse) *shared.AppMcpHTTPResponse {
	encoded, _ := json.Marshal(body)
	return &shared.AppMcpHTTPResponse{
		StatusCode: statusCode,
		Headers: map[string]string{
			"cache-control": "no-store",
			"content-type":  "application/json; charset=utf-8",
		},
		BodyText: string(encoded),
	}
}

func stringsTrim(value string) string {
	return strings.TrimSpace(value)
}
